File: codes/jiheon/238_productofarrayexceptself.java
public int[] productExceptSelf(int[] nums) {

		int zeroIdx = -1, mul=1;
        int[] ans = new int[nums.length];

        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == 0) {
                if (zeroIdx != -1) {
                    return ans;
                } else {
                    zeroIdx = i;
                }
            } else {
                mul *= nums[i];
            }
        }
       
        if (zeroIdx == -1) {
            for (int i = 0; i < nums.length; i++) {
                ans[i] = mul / nums[i];
            }
        } else {
            ans[zeroIdx] = mul;
        }
        return ans;
    }

File: codes/jiheon/238_productofarrayexceptself.kt
fun productExceptSelf(nums: IntArray): IntArray {  
    val ans = IntArray(nums.size){1};  
    
    for (i in 1 until nums.size){  
        ans[i] = ans[i-1]*nums[i-1];  
    }  
  
    var tmp = 1;  
    for (i in nums.size-1 downTo 0){  
        ans[i] *=tmp;  
        tmp*=nums[i]  
    }  
  
    return ans;  
}